mm2src/coins/eth/web3_transport/websocket_transport.rs

```diff
@@ -0,0 +1,317 @@
+use futures::channel::oneshot;
+use futures::lock::Mutex as AsyncMutex;
+use futures_ticker::Ticker;
+use futures_util::{FutureExt, SinkExt, StreamExt};
+use instant::Duration;
+use jsonrpc_core::Call;
+use mm2_net::transport::GuiAuthValidationGenerator;
+use std::collections::HashMap;
+use std::sync::{atomic::{AtomicUsize, Ordering},
+                Arc};
+use web3::error::{Error, TransportError};
+use web3::helpers::to_string;
+use web3::{helpers::build_request, RequestId, Transport};
+
+const REQUEST_TIMEOUT_AS_SEC: u64 = 10;
+
+#[derive(Clone, Debug)]
+pub(crate) struct WebsocketTransportNode {
+    pub(crate) uri: http::Uri,
+    pub(crate) gui_auth: bool,
+}
+
+#[derive(Clone, Debug)]
+pub struct WebsocketTransport {
+    request_id: Arc<AtomicUsize>,
+    node: WebsocketTransportNode,
+    event_handlers: Vec<RpcTransportEventHandlerShared>,
+    pub(crate) gui_auth_validation_generator: Option<GuiAuthValidationGenerator>,
+    responses: Arc<SafeMapPtr>,
+    controller_channel: Arc<ControllerChannel>,
+    connection_guard: Arc<AsyncMutex<()>>,
+}
+
+#[derive(Debug)]
+struct ControllerChannel {
+    tx: Arc<AsyncMutex<UnboundedSender<ControllerMessage>>>,
+    rx: Arc<AsyncMutex<UnboundedReceiver<ControllerMessage>>>,
+}
+
+enum ControllerMessage {
+    Request(WsRequest),
+    Close,
+}
+
+#[derive(Debug)]
+struct WsRequest {
+    serialized_request: String,
+    request_id: RequestId,
+    response_notifier: oneshot::Sender<()>,
+}
+
+/// A wrapper type for raw pointers used as a mutable Send & Sync HashMap reference.
+///
+/// Safety notes:
+///
+/// The implemented algorithm for socket request-response is already thread-safe,
+/// so we don't care about race conditions.
+///
+/// As for deallocations, see the `Drop` implementation below.
+#[derive(Debug)]
+struct SafeMapPtr(*mut HashMap<RequestId, Vec<u8>>);
+
+impl Drop for SafeMapPtr {
+    fn drop(&mut self) {
+        // Let the compiler do the job.
+        let _ = unsafe { Box::from_raw(self.0) };
+    }
+}
+
+unsafe impl Send for SafeMapPtr {}
+unsafe impl Sync for SafeMapPtr {}
+
+impl WebsocketTransport {
+    pub(crate) fn with_event_handlers(
+        node: WebsocketTransportNode,
+        event_handlers: Vec<RpcTransportEventHandlerShared>,
+    ) -> Self {
+        let (req_tx, req_rx) = futures::channel::mpsc::unbounded();
+
+        WebsocketTransport {
+            node,
+            event_handlers,
+            responses: Arc::new(SafeMapPtr(Box::into_raw(Default::default()))),
+            request_id: Arc::new(AtomicUsize::new(1)),
+            controller_channel: ControllerChannel {
+                tx: Arc::new(AsyncMutex::new(req_tx)),
+                rx: Arc::new(AsyncMutex::new(req_rx)),
+            }
+            .into(),
+            connection_guard: Arc::new(AsyncMutex::new(())),
+            gui_auth_validation_generator: None,
+        }
+    }
+
+    pub(crate) async fn start_connection_loop(self) {
```

Comment:
This can probably be refactored by adding some local constants and functions, which will make it more readable. You can add the following constants
```rust
const MAX_ATTEMPTS: u32 = 3;
const SLEEP_DURATION: f64 = 1.;
const SIMPLE_REQUEST: &str = r#"{"jsonrpc":"2.0","method":"net_version","params":[],"id": 0 }"#;
```
And the following functions
```rust
async fn establish_connection(node_uri: &str) -> Result<WebSocketStream, ()>;
async fn send_request(wsocket: &mut WebSocketStream, request: String) -> Result<(), ()>;
// A process request function
// A process message function
```

Reply:
> And the following functions
> 
> ```rust
> async fn establish_connection(node_uri: &str) -> Result<WebSocketStream, ()>;
> async fn send_request(wsocket: &mut WebSocketStream, request: String) -> Result<(), ()>;
> // A process request function
> // A process message function
> ```

It's not possible to parse `select!` flows into that many functions as they based on `loop`s and `continue`s. I did some changes to increase the readibility, I hope this helps.